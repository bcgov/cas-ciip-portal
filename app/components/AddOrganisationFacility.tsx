import React, {useState} from 'react';
import {Button, Modal} from 'react-bootstrap';
import JsonSchemaForm, {IChangeEvent} from 'react-jsonschema-form';
import FormFieldTemplate from 'containers/Forms/FormFieldTemplate';
import FormObjectFieldTemplate from 'containers/Forms/FormObjectFieldTemplate';
import addOperatorSchema from 'components/organisation/addOrganisationSchema';
import addFacilitySchema from 'components/facility/addFacilitySchema';

interface Props {
  onAddOrganisation?: (...args: any[]) => void;
  onAddFacility?: (...args: any[]) => void;
  organisationRowId?: number;
  reportingYear?: number;
}

const AddOrganisationFacility: React.FunctionComponent<Props> = props => {
  const {
    onAddOrganisation,
    onAddFacility,
    organisationRowId,
    reportingYear
  } = props;
  const [modalShow, setModalShow] = useState(false);

  const customFormats = {
    'postal-code': /[A-Z]\d[A-Z]\s?\d[A-Z]\d/i,
    duns: /^\d{9}$/
  };

  const saveNewFacility = async (e: IChangeEvent) => {
    const {
      facilityName,
      facilityType,
      bcghgid,
      naicsCode,
      mailingAddress
    } = e.formData;
    const variables = {
      input: {
        facility: {
          facilityName,
          facilityType,
          bcghgid,
          naicsCode,
          ...mailingAddress,
          // TODO: Our facility table currently stores reportingYear as a string. Should it be an int? It's also a string in swrs & init functions
          reportingYear: String(reportingYear),
          organisationId: organisationRowId
        }
      }
    };
    setModalShow(!modalShow);
    onAddFacility(variables);
  };

  const saveNewOrganisation = async (e: IChangeEvent) => {
    const {operatorName, operatorTradeName, duns, mailingAddress} = e.formData;
    const variables = {
      input: {
        operator: {
          operatorName,
          operatorTradeName,
          duns,
          ...mailingAddress
        }
      }
    };
    setModalShow(!modalShow);
    onAddOrganisation(variables);
  };

  return (
    <>
      {onAddFacility ? (
        <Button onClick={() => setModalShow(!modalShow)}>Add Facility +</Button>
      ) : (
        <p id="add-organisation" onClick={() => setModalShow(!modalShow)}>
          I can&apos;t find my organisation
          <style jsx>
            {`
              #add-organisation {
                color: blue;
              }
              #add-organisation:hover {
                text-decoration: underline;
                cursor: pointer;
              }
            `}
          </style>
        </p>
      )}
      <Modal
        centered
        size="xl"
        show={modalShow}
        onHide={() => setModalShow(false)}
      >
        {onAddFacility ? (
          <Modal.Header>Add Facility</Modal.Header>
        ) : (
          <Modal.Header>Add Organisation</Modal.Header>
        )}
        <Modal.Body>
          <JsonSchemaForm
            schema={onAddFacility ? addFacilitySchema : addOperatorSchema}
            customFormats={customFormats}
            showErrorList={false}
            FieldTemplate={FormFieldTemplate}
            ObjectFieldTemplate={FormObjectFieldTemplate}
            onSubmit={onAddFacility ? saveNewFacility : saveNewOrganisation}
          >
            <Button type="submit" variant="primary">
              Save
            </Button>
            <Button variant="danger" onClick={() => setModalShow(!modalShow)}>
              Close
            </Button>
          </JsonSchemaForm>
        </Modal.Body>
      </Modal>
    </>
  );
};

export default AddOrganisationFacility;
