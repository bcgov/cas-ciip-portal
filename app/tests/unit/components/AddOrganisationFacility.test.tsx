import React from 'react';
import {mount} from 'enzyme';
import AddOrganisationFacility from 'components/AddOrganisationFacility';

describe('AddOrganisationFacility Component', () => {
  it('should match the last snapshot', () => {
    const render = mount(
      <AddOrganisationFacility onAddOrganisation={jest.fn()} />
    );
    expect(render).toMatchSnapshot();
  });

  it("should render the 'I cant find my organisation' Link when passed the onAddOrganisation prop", () => {
    const render = mount(
      <AddOrganisationFacility onAddOrganisation={jest.fn()} />
    );
    expect(render.find('p').text()).toBe("I can't find my organisation");
  });

  it('should render the Add Facility Button when passed the onAddFacility prop', () => {
    const render = mount(<AddOrganisationFacility onAddFacility={jest.fn()} />);
    expect(render.find('Button').text()).toBe('Add Facility +');
  });

  it('should open the modal when Add Facility Button is clicked', () => {
    const render = mount(<AddOrganisationFacility onAddFacility={jest.fn()} />);
    render.find('Button').simulate('click');
    expect(render.find('Modal').exists()).toBe(true);
  });
});
