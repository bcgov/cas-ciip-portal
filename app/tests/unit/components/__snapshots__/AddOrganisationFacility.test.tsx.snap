// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddOrganisationFacility Component should match the last snapshot 1`] = `
Array [
  <p
    className="jsx-2386677284"
    id="add-organisation"
    onClick={[Function]}
  >
    I can't find my organisation
  </p>,
  <Bootstrap(Modal)
    centered={true}
    onHide={[Function]}
    show={false}
    size="xl"
  >
    <ForwardRef
      autoFocus={true}
      backdrop={true}
      backdropTransition={[Function]}
      className="modal"
      containerClassName="modal-open"
      enforceFocus={true}
      keyboard={true}
      manager={
        BootstrapModalManager {
          "adjustAndStore": [Function],
          "containers": Array [],
          "data": Array [],
          "handleContainerOverflow": true,
          "hideSiblingNodes": true,
          "modals": Array [],
          "restore": [Function],
          "scrollbarSize": 0,
        }
      }
      onClick={[Function]}
      onEnter={[Function]}
      onEntering={[Function]}
      onExited={[Function]}
      onHide={[Function]}
      onMouseUp={[Function]}
      renderBackdrop={[Function]}
      restoreFocus={true}
      show={false}
      style={Object {}}
      transition={[Function]}
    />
  </Bootstrap(Modal)>,
]
`;
